lighthouse-core/audits/valid-source-maps.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * @license Copyright 2020 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const thirdPartyWeb = require('third-party-web/httparchive-nostats-subset');
+const Audit = require('./audit.js');
+const i18n = require('../lib/i18n/i18n.js');
+
+// TODO: web.dev docs, write description
+const UIStrings = {
+  /** Title of a Lighthouse audit that provides detail on HTTP to HTTPS redirects. This descriptive title is shown to users when HTTP traffic is redirected to HTTPS. */
+  title: 'Page has valid source maps',
+  /** Title of a Lighthouse audit that provides detail on HTTP to HTTPS redirects. This descriptive title is shown to users when HTTP traffic is not redirected to HTTPS. */
+  failureTitle: 'Missing source maps for large first-party JavaScript',
+  /** Description of a Lighthouse audit that tells the user that their JavaScript source maps are invalid or missing. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
+  description: 'Source maps translate your minified code to the original source code which ' +
+    'helps with debugging, and tracking down bugs in production. Consider enabling or updating ' +
+    'source maps during JavaScript bundling to take advantage of this feature. ' +
+    '[Learn more](https://developers.google.com/web/tools/chrome-devtools/javascript/source-maps).',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+const LARGE_JS_BYTE_THRESHOLD = 500 * 1024;
+
+/**
+ * Returns true if the script URL is either:
+ * (1) a known third-party script or
+ * (2) the same as the (redirected) requested Lighthouse URL
+ * @param {string} url
+ * @param {string} finalUrl
+ */
+function isFirstParty(url, finalUrl) {
```

Comment:
how about `isConsideredFirstPartyByLighthouseButThatsImpliedCuzThisIsInLighthouseCodebaseAlready` :P

Reply:
> how about `isConsideredFirstPartyByLighthouseButThatsImpliedCuzThisIsInLighthouseCodebaseAlready` :P

have to be careful, I might like that name :P

I am somewhat serious, though. I don't think it actually is implied by Lighthouse yet, because I don't believe there's anywhere that we expose "first party" as an explicit classification. We've shown data for known third parties (`third-party-summary`), and we've hidden data for known third parties (`legacy-javascript`), but we haven't said "here's your first-party results" anywhere yet.

There will always be cases like "all schools in our district are required to include this script from some company" and that script isn't in `third-party-web` (but could be added in the future). I don't know if there's any benefit to a nuanced explanation of that in the report if we e.g. switch the `Show 3rd-party resources` filter over to `third-party-web` given the lack of bugs filed on the current, very crude `Show 3rd-party resources` categorization, but internally it makes sense for audit authors to have a clear idea of what the methods are returning (not `isThirdParty`/`isFirstParty` but `isInThirdPartyWeb`/`isNotInThirdPartyWeb`)

Reply:
isInThirdPartyWeb/isNotInThirdPartyWeb wfm

I think long term our solution is more like https://github.com/privacycg/first-party-sets and we tell the schools in that district to follow it :)

Reply:
> I think long term our solution is more like https://github.com/privacycg/first-party-sets and we tell the schools in that district to follow it :)

oh wow, that's pretty great

Reply:
One main reason I'm OK with considering anything not in third-party-web first-party is that if it's not there in the httparchive subset we're currently using, that means it's used on ~0.00125% sites or less which at least suggests that the user of that third-party has a much greater ability to push back on whatever that provider is doing even if they're not in direct control of it.

Reply:
+1. the old version of this PR had this fn, but two other PRs had to use it so we extracted it to that lib.

Reply:
Am I using the lib properly? I feel like I did something wrong by calling `thirdPartyWeb.getEntity`